### PR TITLE
Making it explicit

### PR DESCRIPTION
**Pull request by sapioit**
Because we need it!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,19 +9,18 @@ My Semantic Version schema can be represented like `PUBLIC_API.BACKWARDS_INCOMPA
...
-* **BACKWARDS_INCOMPATIBLE**:
-    + _Any Breaking Code Change_ will increment version to `1.1.0.0`.
-*  **PUBLIC_API**:
-    + _Major Code Overhaul_ that impacts a significant number of end points in the _Public Api_ will increment
-        version to `2.0.0.0`.
+* **Release**
+    + is incremented when changing the API, changing the UX, switching to a new recommended version or ending support for previous versions,
+    + version `1.0.0.0` to `2.0.0.0`.
+* **Incompatible**
+    + is incremented when adding, changing or removing code while breaking compatibility with previous versions or changing the UX,
```
Comment (Exadra37) [inline README.md]: I think is better like:

is incremented when adding, changing, removing code or changing the UX while breaking compatibility with previous versions

**Comment by Exadra37** [comment]
Ping:
@colomet 
@bmitch

**Comment by Exadra37** [comment]
Ping also @corysimmons

**Comment by Exadra37** [comment]
And also:
@veksen
@masaka222
@quesar

**Comment by Exadra37** [comment]
@calidion you may be also interested...

**Comment by sapioit** [comment]
So many pings...

**Comment by Exadra37** [comment]
@xBorderie it seems that Explicit Versioning will be of Prestashop interest?

**Comment by calidion** [comment]
@Exadra37 

if minor changes and patches are both compatible.
they can be unified.

```
Product_Version.Break_Changes.Minor_Changes_OR_Patch.
```

no need to add a new column.

**Comment by colomet** [comment]
@calidion the purpose of 4 digit is to make the people to uderstand nothing new there is in the update. Like wordpress 4.7.X, you can update any time you wish the bug fixes release because you know nothing is going to break the site. If sometimes you have a feature and some times a bug, the people can not be sure they could update.

**Comment by calidion** [comment]
`you know nothing is going to break the site` this declare can not be trusted:)

**Comment by colomet** [comment]
I know, but the most of the time, a new feature in Wordress or Drupal ecosystem can break something. A security release or bug, normally don´t do it

**Comment by calidion** [comment]
so as for me,
rules are:
1. trusted no updates until you fully tested.
2. try to keep api compatibile as much as u can.
3. do mannul updates

so it is not important how many digital it is.
No maintainers like to break their apis, but they can not be trusted either.

it is good to deliver bug fixes and and security patches in a compatible manner.
but it still your duty to fully test them before update them.

**Comment by colomet** [comment]
Wordpress allow to create automatic update, plesk allo to have automatic updates. I don´t think we can do it with 3 digits, with 4 (maybe is not secure and a bad idea) but it could be possible

**Comment by Exadra37** [comment]
@calidion

Explicit Versioning means that we want to know when a Release is just for Fixes, Improvements, Breaking Changes or a Major Release. 

You can read more [here](https://github.com/exadra37-versioning/explicit-versioning/blob/master/WHY.md) about the reasons why SemVer is not addressing the Goals that Explicit Versioning is addressing.

If the software you are building doesn't have a good suite of Tests to cover you against breaking changes, than is an issue in the the Software not with the Versioning Specification you decide to use to versioning it.

If you don't like Explicit Versioning you can always opt in for SemVer specification, that is implicit instead of explicit.

Choose the one that best suites the needs of your Software Release process.

**Comment by calidion** [comment]
with no such trusted, the promise of compatibility is meanless to me.
so i don't think semver will work.
it is just constantly broken.

it is not a problem on how version is conducted, it is how to avoid breaks.

so rules on versioning does affect breaks at all.

only being cautious in updates can avoid such breaks.

though that rules seems attractive but they are lies.

@Exadra37 

I would say your idea is still be on the wrong assumption that people can always keep the promise: that minor version changes or patches won't break. 
it will definitely break as far as i know.

so for me such assumption is no true.
so no need to keep that promise as a rule.

your versioning schema is good to keep product versions but still will be broken as time goes.

So I don't think compatibility is not the thing should be trusted and set as a rule, but we should keep that respect and try not to break as much as possible, and don't  encourage updating.

with no such assumption:
1. major version can be product version
2. minor version can be fixes, improves, small features
3. patches can be patches for buges or security issues.

just update after testing.

**Comment by Exadra37** [comment]
@calidion 

Having the `Incompatible` version number allows to explicit declare when a Software Release is incompatible with the previous release, therefore allowing that we can require packages like `1.0.*` and be sure to not get intentional breaking changes.

Regarding unintentional breaking changes, Explicit Versioning can't guarantee by itself that  Software is not Released with breaking changes when increment only the `Compatible` or `Fix` number, that is a job for the Test Suite in the Software Project.

**Comment by calidion** [comment]
still with `1.0.*` Explicit Versioning has the same problem semver having now, which is unstablity.

**Comment by Exadra37** [comment]
@calidion 

With Explicit Versioning the instability from `1.0.*` only comes from Software that doesn't have a good suite of tests.

Not having a Good Suite of Tests is not a problem for any Versioning Specification to solve, is the responsibility of the Software Project to develop and maintain the tests for any release they do from the first line of code. This includes Acceptance, Integration and Unit Tests.

**Comment by sapioit** [comment]
What many people are omitting is that people already use 4 digits with SimVer, anyway. Be it a `1.2.3.4RC5` or `1.2.3.4-5`, they do have that fourth digit usage, but it's usage is NOT explicit.

Also, @calidion , you're missing the point. Rules will always be broken sooner or later, but we're not talking about removing something that's impossible to remove, we're just trying to minimalize the cyclometric complexity in order to reduce the lack of clarity which make people who DO respect the rules uncounsciously break them, thus making the intended standard not be the same standard for everyone.

In other words, we're only setting some guidlines (which can and WILL be broken, at times) and we're trying to make things not be as vague as they are (because that makes the standard differ from people to people and from project to project).

Also, I think we should all share this with other projects and people (maybe also social media), so as to attract more people that are interested in this.

**Comment by calidion** [comment]
@colomet 

here is a rule: 
"make it work, make it right, make it fast" 
which is still working well.

bad rules are surely broken, but good ones are not.

there are more rules in software that are never broken.

I won't think it is valuable to set rules for breaking.

software developement is some what unpredictable, a restrict rules is of no use.

when software programs are needed, it is out of question if they breaks or not.

when they are not needed, it is useless even if you provide stable and fully tested programs or packages.

compatibility is in effect not a problem, let alone the version?

so versions are in effect only useful to signal that software programs are changed. Big or small changes or just critical fixes.

File: README.md (hunk 1 above)
Comment (Exadra37) [reply README.md]: @sapioit can we discuss this so that we can proceed with the Pull Request?

**Comment by corysimmons** [comment]
Can someone find out if npm cares about appended stuff?

Like `99.99.99-v2` could serve the same purpose but be backwards compatible.

**Comment by sapioit** [comment]
>"make it work, make it right, make it fast"
>...
>bad rules are surely broken, but good ones are not.

Said every rapist ever...
And what makes you think people don't break that rule, either? How many projects are started and not finished? Do they work? Well, since they're unfinished, they don't work, are probably not made right and are not necessarily made fast, either. **So yeah, your argument is invalid.**

Because people also need to know when it's safe to update their software, since when working with banking or health, or military, or safety, or other dangerous things, you definitely want to know if it's okay to update. And that applies both to the users and the developers (who are users of the software they write, by having to test it to make sure it works).

So yeah, we could argue about this for centuries, but let's better focus on making this a standard, rather than arguing about it's marginal usefulness to someone who will not use it.

@corysimmons That's how people already use it. And yes, whether you use a dot or a dash, so long as it's the same logic behind, it's all good.

**Comment by Exadra37** [comment]
>Like 99.99.99-v2 could serve the same purpose but be backwards compatible.

Everything after the dash is not part of the specification.

After the dash is to mark the release as:

* Alpha Release
* Release Candidate
* Beta Release
* Any other identifier you workflow you process requires.

After Dash can't be used to apply the Explicit Versioning specification.

I will address this in another issue, that will clarify the optional identifiers.

File: README.md (hunk 1 above)
Comment (sapioit) [reply README.md]: Yes, it is alright. It makes more sense to have it that way.
Sorry, in the sea of comments I forgot to look up here...

**Review by sapioit — COMMENTED** [review]
I hope this fixes it...

**Comment by calidion** [comment]
@sapioit 

when you think that assumption is right for you, then semver is the right choice.
no need for this versioning schema.
because you can add more digitals.

like: product_ver.company_ver.software_ver.upgrade_ver.fix_ver

software versioning should only reflect software's own changes.

so major.minor.patch is enougth.

As for me. I am not upset with the major.minor.patch format.
but the underlining assumption and auto upgrade.

the auto upgrade make pacakges unstable, while the promising of the compatibility drived the wrong decision of auto upgrade, then result in version booming.

if there is no such promise and make small breaks acceptable in versioning then everything will work well and in harmoney.

the version booming problem is due to the wrong assumption, not the major.minor.patch format is wrong.

**Comment by corysimmons** [comment]
Can someone update me? I stopped reading after I proposed. I just wanted it to tempt marketing-centric teams from using semver wrong.

Not sure where you guys took this.

**Comment by calidion** [comment]
I have made a chinese version guide to prevent such unstability.

https://github.com/calidion/nodever

i have fully expressed my idea here:
https://github.com/npm/npm/issues/16432

File: README.md [hunk 2]
```diff
@@ -1,27 +1,26 @@
...
-    + _New Features_ will increment version to `1.0.1.0`.
-    + _Refracting Code_ that do not impact _Public Api_ will increment version to `1.0.1.0`.
-    + _Deprecating Code_ that do not impact _Public Api_ will increment version to `1.0.1.0`.
-    + _Removing Code_ that do not impact _Public Api_ will increment version to `1.0.1.0`.
-* **BACKWARDS_INCOMPATIBLE**:
-    + _Any Breaking Code Change_ will increment version to `1.1.0.0`.
-*  **PUBLIC_API**:
-    + _Major Code Overhaul_ that impacts a significant number of end points in the _Public Api_ will increment
-        version to `2.0.0.0`.
+* **Release**
```
Comment (calidion) [inline README.md:12]: Is 0.1.0.0 not a release?

Comment (Exadra37) [reply README.md:12]: Is a development release that is Incompatible with previous version.

But the Readme starts with a production Release `1.0.0.0`.

Comment (calidion) [reply README.md:12]: if i released a version 0.1.0.0. can it be used as a production release?
if it can, then what is the use of this release digital?
or if it cannot, why should i give it a version at all?

Comment (Exadra37) [reply README.md:12]: The most left number must be incremented to `1` only when your Software is ready to be used in Production.

Explicit Versioning specification or any other Versioning specification can't enforce that Developers shouldn't use a Developer Release in their Production Software, but Package Manager can/should enforce it.

Resuming, enforcing not using Development Releases in Production is out of scope for any Versioning Specification schema, it can be only recommended as a best practice.

**Comment by sapioit** [comment]
>As for me. I am not upset with the major.minor.patch format.
but the underlining assumption and auto upgrade.
>
>the auto upgrade make pacakges unstable, while the promising of the compatibility drived the wrong decision of auto upgrade, then result in version booming.

While that might be the case, there is also the need for the software that deals the auto-updates to make backups before updating (configurable) and store them (by default) for a month (configurable). Because that's not only a matter of whether the files update well, but also a question of the integrity of the server and other possible security gaps.

But we're not talking about things package managers, which can be malfunctioning themselves (as in incomplete, bugged or right code for a broken logic), but a set of guidlines for the developers.

Realistically speaking, nothing stops you from releasing every Release version as it's own package (i.e. Github 1, Github 2, Github 3, or Github A, Github B, Github C, ..., Github Z, Github AA) and use the rest of the digits for the versioning of those recommended branches.

File: README.md (hunk 2 above)
Comment (Exadra37) [reply README.md:12]: @sapioit 

Please can you squash the commits to only 1 commit, so that we can keep a clean git history without repeat committed messages.

**Comment by sapioit** [comment]
So... are you going to merge this or not? We can move to a different issue to continue discussing, you know?
But even with the title changed, the README still says **Semantic Versioning**...

**Review by Exadra37 — CHANGES_REQUESTED** [review]

**Comment by sapioit** [comment]
Seems like you did that already. 
Thanks!

**Comment by Exadra37** [comment]
Yeah, Github allows us to Squash commits on Merge ;)
